subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,958 @@
+					USB_AUDIO_USB_STREAMING),	  \
+	.feature_unit_1 = INIT_FEATURE_UNIT(dev##_HP, i, id+4, id+3),	  \
+	.output_terminal_1 = INIT_OUT_TERMINAL(id+5, id+4,		  \
+						USB_AUDIO_IO_HEADSET),	  \
+	.as_interface_alt_0_0 = INIT_STD_IF(USB_AUDIO_AUDIOSTREAMING,	  \
+						1, 0, 0),		  \
+	.as_interface_alt_0_1 = INIT_STD_IF(USB_AUDIO_AUDIOSTREAMING,	  \
+						1, 1, 1),		  \
+		.as_cs_interface_0 = INIT_AS_GENERAL(id+2),		  \
+		.format_0 = INIT_AS_FORMAT_I(CH_CNT(dev##_MIC, i),	  \
+					     GET_RES(dev##_MIC, i)),	  \
+		.std_ep_desc_0 = INIT_STD_AS_AD_EP(dev##_MIC, i,	  \
+						   AUTO_EP_IN),		  \
+		.cs_ep_desc_0 = INIT_CS_AS_AD_EP,			  \
+	.as_interface_alt_1_0 = INIT_STD_IF(USB_AUDIO_AUDIOSTREAMING,	  \
+						2, 0, 0),		  \
+	.as_interface_alt_1_1 = INIT_STD_IF(USB_AUDIO_AUDIOSTREAMING,	  \
+						2, 1, 1),		  \
+		.as_cs_interface_1 = INIT_AS_GENERAL(id+3),		  \
+		.format_1 = INIT_AS_FORMAT_I(CH_CNT(dev##_HP, i),	  \
+					     GET_RES(dev##_HP, i)),	  \
+		.std_ep_desc_1 = INIT_STD_AS_AD_EP(dev##_HP, i,		  \
+						   AUTO_EP_OUT),	  \
+		.cs_ep_desc_1 = INIT_CS_AS_AD_EP,			  \
+};									  \
+static struct usb_ep_cfg_data dev##_usb_audio_ep_data_##i[] = {		  \
+	INIT_EP_DATA(usb_transfer_ep_callback, AUTO_EP_IN),		  \
+	INIT_EP_DATA(audio_receive_cb, AUTO_EP_OUT),			  \
+}
+
+#define DEFINE_AUDIO_DEV_DATA(dev, i, __out_pool, __in_pool_size)     \
+	static u8_t dev##_controls_##i[FEATURES_SIZE(dev, i)] = {0};  \
+	static struct usb_audio_dev_data_t dev##_audio_dev_data_##i = \
+		{ .pool = __out_pool,				      \
+		  .in_frame_size = __in_pool_size,		      \
+		  .controls = {dev##_controls_##i, NULL},	      \
+		  .ch_num = {(CH_CNT(dev, i) + 1), 0}		      \
+		}
+
+#define DEFINE_AUDIO_DEV_DATA_BIDIR(dev, i, __out_pool, __in_pool_size)	   \
+	static u8_t dev##_controls0_##i[FEATURES_SIZE(dev##_MIC, i)] = {0};\
+	static u8_t dev##_controls1_##i[FEATURES_SIZE(dev##_HP, i)] = {0}; \
+	static struct usb_audio_dev_data_t dev##_audio_dev_data_##i =	   \
+		{ .pool = __out_pool,					   \
+		  .in_frame_size = __in_pool_size,			   \
+		  .controls = {dev##_controls0_##i, dev##_controls1_##i},  \
+		  .ch_num = {(CH_CNT(dev##_MIC, i) + 1),		   \
+			     (CH_CNT(dev##_HP, i) + 1)}			   \
+		}
+
+/**
+ * Helper function for getting channel number directly from the
+ * feature unit descriptor.
+ */
+static u8_t get_num_of_channels(struct feature_unit_descriptor *fu)
+{
+	return (fu->bLength - FU_FIXED_ELEMS_SIZE)/sizeof(u16_t);
+}
+
+/**
+ * Helper function for getting supported controls directly from
+ * the feature unit descriptor.
+ */
+static u16_t get_controls(struct feature_unit_descriptor *fu)
+{
+	return *(u16_t *)((u8_t *)fu + BMA_CONTROLS_OFFSET);
+}
+
+/**
+ * Helper function for getting the device streaming direction
+ */
+static enum usb_audio_direction get_fu_dir(struct feature_unit_descriptor *fu)
+{
+	struct output_terminal_descriptor *ot  =
+	(struct output_terminal_descriptor *)((u8_t *)fu + fu->bLength);
```

Comment:
Imperfect indentation. And `const`s missing (in the function parameter as well).